runtime/ts/plan/plan-consumer.ts

```diff
@@ -0,0 +1,131 @@
+/**
+ * @license
+ * Copyright (c) 2018 Google Inc. All rights reserved.
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ * Code distributed by Google as part of this project is also
+ * subject to an additional IP rights grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+import {assert} from '../../../platform/assert-web.js';
+import {Arc} from '../arc';
+import {Recipe} from '../recipe/recipe';
+import {PlanningResult} from './planning-result';
+import {StorageProviderBase} from '../storage/storage-provider-base';
+import {SuggestionComposer} from '../../suggestion-composer.js';
+
+type Callback = ({}) => void;
+
+export class PlanConsumer {
+  arc: Arc;
+  result: PlanningResult;
+  store: StorageProviderBase;
+  suggestFilter: {};
+  plansChangeCallbacks: Callback[] = [];
+  suggestionsChangeCallbacks: Callback[] = [];
+  storeCallback: Callback;
+  suggestionComposer: SuggestionComposer|null = null;
+
+  constructor(arc: Arc, store: StorageProviderBase) {
+    assert(arc, 'arc cannot be null');
+    assert(store, 'store cannot be null');
+    this.arc = arc;
+    this.result = new PlanningResult(arc);
+    this.store = store;
+    this.suggestFilter = {showAll: false};
+    this.plansChangeCallbacks = [];
+    this.suggestionsChangeCallbacks = [];
+
+    this.storeCallback = () => this.onStoreChanged();
+    this.store.on('change', this.storeCallback, this);
+
+    // create suggestion composer and register for a callback.
+    const composer = arc.pec.slotComposer;
+    if (composer) {
+      if (composer.findContextById('rootslotid-suggestions')) {
+        this.suggestionComposer = new SuggestionComposer(composer);
+        this.registerSuggestChangedCallback(
+            (suggestions) => this.suggestionComposer.setSuggestions(suggestions));
+      }
+    }
+  }
+
+  registerPlansChangedCallback(callback) { this.plansChangeCallbacks.push(callback); }
+  registerSuggestChangedCallback(callback) { this.suggestionsChangeCallbacks.push(callback); }
+
+  setSuggestFilter(showAll, search) {
+    assert(!showAll || !search);
+    if (this.suggestFilter['showAll'] === showAll && this.suggestFilter['search'] === search) {
+      return;
+    }
+    const previousSuggestions = this.getCurrentSuggestions();
+    this.suggestFilter = {showAll, search};
```

Comment:
Optional code style note: move line 63 above line 62, this is only a readability thing; as a reader the unusual order of operations tripped me up

Reply:
changing `this.suggestFilter` affects `this.getCurrentSuggestions()` results, that's why i need to get the `previousSuggestions` before I change the filter.